Add  unifi sensor for number of clients to device

kimdv opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
This PR aims to add a sensor to a Unifi device, so it's possible to see how many clients are connected to it. 

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

kimdv committed:
Add sensor for number of clients to device

kimdv commented:
This PR is created as a draft, just to see if I'm on the right path.

Kane610 commented:
Yes. Go on 👍🏻

Kane610 commented:
I've published aiounifi v78 https://github.com/Kane610/aiounifi/compare/v77...v78

Kane610 commented:
This one is nearly complete, maybe fix this simple one in between the more complex PRs?

Kane610 commented on homeassistant/components/unifi/sensor.py:
```diff
@@ -108,6 +108,27 @@ def async_wlan_client_value_fn(hub: UnifiHub, wlan: Wlan) -> int:
+                )
+                or (client.access_point_mac == "" and client.switch_mac == device.mac)
+            )
+            and dt_util.utcnow() - dt_util.utc_from_timestamp(client.last_seen or 0)
+            < hub.config.option_detection_time
```
Im not sure if we should consider this data for this sensor

kimdv replied:
So my initial thought is that I want to know how many devices is on the network now.
If I understand it correctly, when removing the number will only increase when devices join the network right ?

Kane610 replied:
It would presumably just take a bit longer before the number changes

kimdv replied:
Just to be clear and I fully understand it 
 will the number always increase and never decrease then?

Kane610 replied:
Remove it and test it out :)

I would assume it would decrease (but maybe in a lower fashion) as well but if it doesnt we keep it in

kimdv replied:
Removed ti

Kane610 replied:
So you tester it and it still works?

Kane610 replied:
Sorry for back and forth here, I saw that the same thing was done on the wlan client side, so lets keep it as it was considering the last seen as well.

Kane610 reviewed (commented):
Lets just discuss this point first

kimdv committed:
Remove last_seen

dknowles2 committed:
Bump pydrawise to 2024.6.3 (#118977)

cdce8p committed:
Import Generator from typing_extensions (3) (#118990)

epenet committed:
Move mock_bluetooth* fixtures to decorator (#118846)

cdce8p committed:
Import Generator from typing_extensions (5) (#118993)

cdce8p committed:
Import Generator from typing_extensions (4) (#118992)

epenet committed:
Improve type hints in apple_tv tests (#118980)

cdce8p committed:
Increment ci cache version (#118998)

bdraco committed:
Only entity verify state writable once after success unless hass is missing (#118896)

zxdavb committed:
Move evohome's API broker to the coordinator module (#118565)

* move Broker to coordinator module

* mypy tweak

* mypy

Commit:
Set username as entry title in Bring integration (#118974)

Set username as entry title

sdb9696 committed:
Fully mock the ring_doorbell api and remove requests_mock (#113140)

* Fully mock ring_doorbell library

* Add comments and docstrings

* Simplify devices_mocks and fake RingDevices

* Update post review

* Consolidate device filtering in conftest

* Fix ruff lambda assignment failure

* Fix ruff check fail

* Update post review

gedaliahknizhnik committed:
Add extra sensor to the Jewish Calendar integration (#116734)

* Added sensor for always three star tzeit

* Changed sensor name

* Change sensor name

synesthesiam committed:
Prioritize literal text with name slots in sentence matching (#118900)

Prioritize literal text with name slots

Kane610 committed:
Use fixtures in UniFi diagnostics tests (#118905)

Kane610 committed:
Use fixtures in UniFi image tests (#118887)

Jordi1990 committed:
Add Aquacell integration (#117117)

* Initial commit

* Support changed API
Change sensor entity descriptions

* Fix sensor not handling coordinator update

* Implement re-authentication flow and handle token expiry

* Bump aioaquacell

* Bump aioaquacell

* Cleanup and initial tests

* Fixes for config flow tests

* Cleanup

* Fixes

* Formatted

* Use config entry runtime
Use icon translations
Removed reauth
Removed last updated sensor
Changed lid in place to binary sensor
Cleanup

* Remove reauth strings

* Removed binary_sensor platform
Fixed sensors not updating properly

* Remove reauth tests
Bump aioaquacell

* Moved softener property to entity class
Inlined validate_input method
Renaming of entities
Do a single async_add_entities call to add all entities
Reduced code in try blocks

* Made tests parameterized and use test fixture for api
Cleaned up unused code
Removed traces of reauth

* Add check if refresh token is expired
Add tests

* Add missing unique_id to config entry mock
Inlined _update_config_entry_refresh_token method
Fix incorrect test method name and comment

* Add snapshot test
Changed WiFi level to WiFi strength

* Bump aioaquacell to 0.1.7

* Move test_coordinator tests to test_init
Add test for duplicate config entry

farmio committed:
Fix KNX `climate.set_hvac_mode` not turning `on` (#119012)

engrbm87 committed:
Use torrent id to identify torrents that should trigger events (#118897)

elupus committed:
Update gardena library to 1.4.2 (#119010)

synesthesiam committed:
Fix exposure checks on some intents (#118988)

* Check exposure in climate intent

* Check exposure in todo list

* Check exposure for weather

* Check exposure in humidity intents

* Add extra checks to weather tests

* Add more checks to todo intent test

* Move climate intents to async_match_targets

* Update test_intent.py

* Update test_intent.py

* Remove patch

bieniu committed:
Add the missing humidity value to the Accuweather daily forecast (#119013)

Add the missing humidity value to the daily forecast

Co-authored-by: Maciej Bieniek <478555+bieniu@users.noreply.github.com>

cdce8p committed:
Fix Generator annotations in tests (2) (#119019)

cdce8p committed:
Fix Generator annotations in tests (1) (#119018)

epenet committed:
Improve type hints in airnow tests (#119038)

bdraco committed:
Fix remember_the_milk calling configurator async api from the wrong thread (#119029)

joostlek committed:
Fix unit of measurement for airgradient sensor (#118981)

joostlek committed:
Fix AirGradient name (#119046)

epenet committed:
Remove unused request fixtures (#119044)

gjohansson-ST committed:
Ignore deprecation warning in python-holidays (#119007)

bdraco committed:
Remove isal from after_dependencies in http (#119000)

thecode committed:
Hold connection lock in Shelly RPC reconnect (#119009)

lnx85 committed:
Ecovacs get_positions service (#118572)

Co-authored-by: Robert Resch <robert@resch.dev>

CFenner committed:
Add photovoltaic sensors to ViCare integration (#113664)

* Add photovoltaic sensors

* Update strings.json

* Apply suggestions from code review

* change uom for daily sensor

epenet committed:
Improve type hints in zamg tests (#119042)

epenet committed:
Improve type hints in motionblinds_ble tests (#119049)

epenet committed:
Improve type hints in kaleidescape tests (#119040)

epenet committed:
Improve type hints in azure_event_hub tests (#119047)

Co-authored-by: Marc Mueller <30130371+cdce8p@users.noreply.github.com>

mdegat01 committed:
Always have addon url in detached_addon_missing (#119011)

gjohansson-ST committed:
Bump babel to 2.15.0 (#119006)

dependabot[bot] committed:
Bump dawidd6/action-download-artifact from 4 to 5 (#118851)

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

huyuwei1996 committed:
Enable retrieving sensor data from WoHub2 device and update pySwitchbot to 0.47.2 (#118567)

jeeftor committed:
Align weatherflow_cloud weather conditions with Home Assistant supported conditions (#114497)

* WeatherFlow Cloud changing icon mapping specificaly for Colorado

* changing name to state_map

cdce8p committed:
Add type ignore comments (#119052)

epenet committed:
Improve type hints in homekit_controller tests (#119053)

epenet committed:
Remove unused caplog fixtures in tests (#119056)

epenet committed:
Improve type hints in config tests (#119055)

epenet committed:
Add type hints to current_request_with_host in tests (#119054)

epenet committed:
Add type hint to mock_async_zeroconf in test fixtures (#119057)

joostlek committed:
Bump aiowithings to 3.0.1 (#118854)

Kane610 committed:
Use fixtures in UniFi sensor tests (#118921)

pavoni committed:
Add mute_toggle to roon volume events (#114171)

Add mute_toggle event.

yuxincs committed:
Refine sensor descriptions for APCUPSD (#114137)

* Refine sensor descriptions for APCUPSD

* Add device class for cumonbatt

* Add UoM to STESTI and TIMELEFT

* Remove device class for STESTI

epenet committed:
Move mock_async_zeroconf to decorator in homekit tests (#119060)

Darsstar committed:
Bump solax from 3.1.0 to 3.1.1 (#118888)

epenet committed:
Improve type hints in arcam_fmj tests (#119072)

epenet committed:
Improve type hints in airvisual test fixtures (#119079)

joostlek committed:
Fix Azure Data Explorer strings (#119067)

paulusbrand committed:
Add template Base64 decode encoding parameter (#116603)

Co-authored-by: Robert Resch <robert@resch.dev>

epenet committed:
Improve type hints in assist_pipeline tests (#119066)

epenet committed:
Fix incorrect type hints in azure_data_explorer tests (#119065)

bdraco committed:
Fix refactoring error in snmp switch (#119028)

OzGav committed:
Add previous track intent (#113222)

* add previous track intent

* add stop and clear playlist

* Remove clear_playlist and stop

* Remove clear_playlist and stop

* Use extra constraints

---------

Co-authored-by: Michael Hansen <mike@rhasspy.org>

bdraco committed:
Bump home-assistant-bluetooth to 1.12.1 (#119026)

epenet committed:
Move mock_async_zeroconf to decorator in zeroconf tests (#119063)

tronikos committed:
Bump google-generativeai to 0.6.0 (#119062)

epenet committed:
Move mock_zeroconf to decorator in tests (#119061)

emontnemery committed:
Add support for common references in strings.json (#118783)

* Add support for common references in strings.json

* Update tests

Kane610 committed:
Use fixtures in UniFi device tracker tests (#118912)

cdce8p committed:
Update typing-extensions to 4.12.2 (#119098)

Bre77 committed:
Improve the reliability of tests in Tessie (#118596)

tronikos committed:
Catch GoogleAPICallError in Google Generative AI (#119118)

rwalker777 committed:
Add Tuya based bluetooth lights (#119103)

tronikos committed:
Properly handle escaped unicode characters passed to tools in Google Generative AI (#119117)

t0bst4r committed:
Add intelligent language matching for Google Assistant SDK Agents (#112600)

Co-authored-by: Erik Montnemery <erik@montnemery.com>

thecode committed:
Bump aioshelly to 10.0.1 (#119123)

rokam committed:
Bump sunweg to 3.0.1 (#118435)

farmio committed:
Require KNX boolean service descriptor selectors (#118597)

joostlek committed:
Bump aiowaqi to 3.1.0 (#119124)

joostlek committed:
Add support for segmental measurements in Withings (#119126)

joostlek committed:
Improve withings diagnostics (#119128)

bieniu committed:
Use translation placeholders in AccuWeather (#118760)

* Use translation placeholder

* Update test snapshot

---------

Co-authored-by: Maciej Bieniek <478555+bieniu@users.noreply.github.com>

bieniu committed:
Add icon translations to AccuWeather (#119134)

Co-authored-by: Maciej Bieniek <478555+bieniu@users.noreply.github.com>

rikroe committed:
Change BMW select and sensor enums to lowercase (#118751)

Co-authored-by: Richard <rikroe@users.noreply.github.com>

epenet committed:
Add type hints for service_calls fixture in pylint plugin (#118356)

NECH2004 committed:
Refactor Zeversolar init tests (#118551)

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

rikroe committed:
Adjust BMW enum sensors translations (#118754)

Co-authored-by: Richard <rikroe@users.noreply.github.com>

raman325 committed:
Add new security keys to zwave_js config flow (#115835)

frenck committed:
Update Home Assistant base image to 2024.06.0 (#119147)

TomBrien committed:
Fix workday timezone (#119148)

balloob committed:
Ensure intent tools have safe names (#119144)

Kane610 committed:
Fix failing UniFi tests related to utcnow (#119131)

* test

* Fix missed test

autinerd committed:
Enable Ruff PT012 (#113957)

xyzroe committed:
Add ZHA XZG firmware discovery (#116828)

autinerd committed:
Use runtime_data for enigma2 (#119154)

* Use runtime_data for enigma2

* Update __init__.py

mib1185 committed:
Bump py-synologydsm-api to 2.4.4 (#119156)

bump py-synologydsm-api to 2.4.4

Kane610 committed:
Remove old UniFi test infrastructure (#119160)

Clean up hub

Guy293 committed:
Fix Tami4 component breaking API changes (#119158)

* fix tami4 api breaking changes

* fix tests

tronikos committed:
Make supported_features of manual alarm_control_panel configurable (#119122)

bdraco committed:
Ensure multiple executions of a restart automation in the same event loop iteration are allowed (#119100)

* Add test for restarting automation

related issue #119097

* fix

* add a delay since restart is an infinite loop

* tests

mib1185 committed:
Use more conservative timeout values in Synology DSM (#119169)

use ClientTimeout object

tkdrob committed:
Remove Netgear LTE yaml import (#119180)

Remove Netgear LTE yaml config

tronikos committed:
Bump opower to 0.4.7 (#119183)

jschlyter committed:
Add myself as codeowner for `amazon_polly` (#119189)

autinerd committed:
Add fallback to entry_id when no mac address is retrieved in enigma2 (#119185)

thecode committed:
Handle Shelly BLE errors during connect and disconnect (#119174)

etnoy committed:
Fix sia custom bypass arming in night mode (#119168)

Kane610 committed:
Rewrite the UniFi button entity tests (#118771)

bdraco committed:
Migrate august to use yalexs 5.2.0 (#119178)

joostlek committed:
Remove Harmony switches (#119206)

jbouwh committed:
Bump incomfort backend library to v0.6.0 (#119207)

Kane610 committed:
Rework UniFi tests to not use runtime data (#119202)

joostlek committed:
Add config flow to MPD (#117907)

emontnemery committed:
Improve demo config flow and add tests (#118481)

Co-authored-by: epenet <6771947+epenet@users.noreply.github.com>

joostlek committed:
Create DWD device with unique_id instead of entry_id (#116498)

Co-authored-by: Matthias Alphart <farmio@alphart.net>

jbouwh committed:
Bump incomfort backend client to v0.6.1 (#119209)

eifinger committed:
Add reconfigure step for google_travel_time (#115178)

* Add reconfigure step for google_travel_time

* Do not allow to change name

* Duplicate tests for reconfigure

* Use link for description in strings.json

* Try except else

* Extend existing config flow tests

gjohansson-ST committed:
Fix envisalink alarm (#119212)

elupus committed:
Update links between config entry and device on sensor change in integral (#119213)

Bre77 committed:
Add cabin overheat protection entity to Teslemetry (#118449)

* test_cabin_overheat_protection

* Fix snapshot

* Translate error

* Review Feedback

bdraco committed:
Use a listcomp for history results (#119188)

bdraco committed:
Reduce overhead to reduce statistics (#119187)

jbouwh committed:
Fix - Remove unneeded assert in teslemetry test (#119219)

Remove unneded assert in teslemetry test

JakeMartin-ICL committed:
Bump monzopy to 1.3.0 (#119225)

epenet committed:
Improve type hints in ambient_station tests (#119230)

epenet committed:
Improve type hints in amberelectric tests (#119229)

Kane610 committed:
Fixture cleanup in UniFi tests (#119227)

* Make sure all mock_device_registry are used with usefixtuers

* Make sure to use name with fixtures and rename functions to start with fixture_

* Streamline config_entry_setup

* Type all *_payload

* Mark @pytest.mark.usefixtures("mock_default_requests")

* Clean up unnecessary newlines after docstring

dknowles2 committed:
Bump pyschlage to 2024.6.0 (#119233)

starkillerOG committed:
Bump reolink-aio to 0.9.2 (#119236)

Commit:
Bump python-roborock to 2.3.0 (#119228)

starkillerOG committed:
Add Reolink manual record switch (#119232)

Add manual record switch

bdraco committed:
Switch unifiprotect lib to use uiprotect (#119243)

abmantis committed:
Fix wrong arg name in Idasen Desk config flow (#119247)

jbouwh committed:
Add boiler entity state translations for incomfort water_heater entities (#119211)

elupus committed:
Always provide a currentArmLevel in Google assistant (#119238)

Bre77 committed:
Add model to Teslemetry Wall Connectors (#119251)

frenck committed:
Temporary pin CI to Python 3.12.3 (#119261)

bdraco committed:
Fix climate on/off in nexia (#119254)

allenporter committed:
Fix nest to cancel event listener on config entry unload (#119257)

wittypluck committed:
Fix Glances v4 network and container issues (glances-api 0.8.0) (#119226)

eifinger committed:
google_travel_time: Merge user_input validation (#119221)

allenporter committed:
Bump google-nest-sdm to 4.0.5 (#119255)

bdraco committed:
Bump uiprotect to 0.4.0 (#119256)

epenet committed:
Improve type hints in blackbird tests (#119262)

angelnu committed:
Fixes crashes when receiving malformed decoded payloads (#119216)

Co-authored-by: Jan Bouwhuis <jbouwh@users.noreply.github.com>

dgomes committed:
Moves V2C from hass.data to config_entry.runtime_data (#119165)

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

adrum committed:
Fix control 4 on os 2 (#119104)

epenet committed:
Improve type hints in blueprint tests (#119263)

epenet committed:
Improve type hints in axis tests (#119260)

* Improve type hints in axis tests

* A couple more

* One more

* Improve light

* Improve hub

* Improve config-flow

epenet committed:
Improve type hints in camera tests (#119264)

epenet committed:
Use device_registry fixture in webostv tests (#119269)

epenet committed:
Use service_calls fixture in bthome tests (#119268)

jbouwh committed:
Improve incomfort boiler state strings (#119270)

epenet committed:
Use service_calls fixture in scaffold (#119266)

epenet committed:
Use relative imports in tests [a-i] (#119280)

epenet committed:
Use service_calls fixture in arcam_fmj tests (#119274)

epenet committed:
Fix root-import pylint warning in dlna_dmr tests (#119286)

epenet committed:
Use relative imports in tests [j-r] (#119282)

epenet committed:
Use relative imports in tests [s-z] (#119283)

epenet committed:
Avoid references to websocket_api.const in core and other components (#119285)

Sebclem committed:
Change qBittorrent lib to qbittorrentapi (#113394)

* Change qBittorrent lib to qbittorrentapi

* Fix tests

* Convert qbittorrent service to new lib

* Add missing translation key

* Catch APIConnectionError in service call

* Replace type ignore by Any typing

* Remove last type: ignore

* Use lib type for torrent_filter

* Change import format

* Fix remaining Any type

---------

Co-authored-by: Erik Montnemery <erik@montnemery.com>

frenck committed:
Fix persistence on OpenWeatherMap raised repair issue (#119289)

epenet committed:
Don't run tests if lint-ruff-format fails (#119291)

epenet committed:
Use service_calls fixture in helper tests (#119275)

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

epenet committed:
Use Registry fixture in zwave_js tests (#119277)

epenet committed:
Fix root-import pylint warning in components (#119294)

* Fix root-import pylint warning in components

* Adjust

* Adjust

epenet committed:
Fix namespace-import pylint warning in components (#119299)

chammp committed:
Add code_format_template to template locks (#106947)

* Add code_format to template locks

* Replace code_format with code_format_template

* Add test case for template eval to None

* Apply suggestion to not call super()

Co-authored-by: Erik Montnemery <erik@montnemery.com>

* Add more negative tests

* Handle template render errors

* Better error message

* Add custom test lock config for code format

* Add type hints from upstream

---------

Co-authored-by: Erik Montnemery <erik@montnemery.com>
Co-authored-by: Franck Nijhof <git@frenck.dev>

wfltaylor committed:
Add support for animal detection in unifiprotect (#116290)

Co-authored-by: J. Nick Koston <nick@koston.org>

tronikos committed:
Log errors in Intent.async_handle (#119182)

* Log errors in Intent.async_handle

* log exception stack trace

* Update homeassistant/helpers/intent.py

---------

Co-authored-by: Paulus Schoutsen <paulus@home-assistant.io>

chemelli74 committed:
bump aiobotocore to 2.13.0 (#119297)

bump aiobotocore

bdraco committed:
Bump uiprotect to 0.4.1 (#119308)

jbouwh committed:
Add HVACAction to incomfort climate devices (#119315)

* Add HVACAction to incomfort climate devices

* Use IDLE state when not heating

jbouwh committed:
Fix incomfort water heater return translated fault code (#119311)

zweckj committed:
Migrate lamarzocco to lmcloud 1.1 (#113935)

* migrate to 1.1

* bump to 1.1.1

* fix newlines docstring

* cleanup entity_description fns

* strict generics

* restructure import

* tweaks to generics

* tweaks to generics

* removed exceptions

* move initialization, websocket clean shutdown

* get rid of duplicate entry addign

* bump lmcloud

* re-add calendar, auto on/off switches

* use asdict for diagnostics

* change number generator

* use name as entry title

* also migrate title

* don't migrate title

* remove generics for now

* satisfy mypy

* add s

* adapt

* migrate entry.runtime_data

* remove auto/onoff

* add issue on wrong gw firmware

* silence mypy

* remove breaks in ha version

* parametrize issue test

* Update update.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* Update test_config_flow.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

* regen snapshots

* mapping steam level

* remove commented code

* fix typo

* coderabbitai availability tweak

* remove microsecond moving

* additonal schedule for coverage

* be more specific on date offset

* keep mappings the same

* config_entry imports sharpened

* remove unneccessary testcase, clenup date moving

* remove superfluous calendar testcase from diag

* guard against future version downgrade

* use new entry for downgrade test

* switch to lmcloud 1.1.11

* revert runtimedata

* revert runtimedata

* version to helper

* conistent Generator

* generator from typing_extensions

---------

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

jpbede committed:
Add more debug logging to Ping integration (#119318)

epenet committed:
Revert SamsungTV migration (#119234)

bramkragten committed:
Update frontend to 20240610.0 (#119320)

karwosts committed:
Fix statistic_during_period wrongly prioritizing ST statistics over LT (#115291)

* Fix statistic_during_period wrongly prioritizing ST statistics over LT

* comment

* start of a test

* more testcases

* fix sts insertion range

* update from review

* remove unneeded comments

* update logic

* min/mean/max testing

epenet committed:
Improve type hints in registry helper tests (#119302)

bdraco committed:
Improve workday test coverage (#119259)

epenet committed:
Improve type hints in aiohttp_client helper tests (#119300)

cyr-ius committed:
Use runtime_data in Husqvarna Automower (#119309)

osohotwateriot committed:
Add OSO Energy temperature sensors (#119301)

jbouwh committed:
Improve incomfort binary sensors (#119292)

* Improve incomfort binary_sensor, add is_burning, is_pumping and is_tapping

* Update test snapshot

* Use helper for fault code label name

* Update tests

* Remove extra state attribute

* Make default Fault `none` to supprt localization

* Update snapshot

LapsTimeOFF committed:
Fix elgato light color detection (#119177)

Bre77 committed:
Refactor helpers and bump Teslemetry (#119246)

autinerd committed:
Add options flow to enigma2 (#115795)

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

basbruss committed:
Add price service call to Tibber (#117366)

Co-authored-by: Daniel Hjelseth Høyer <mail@dahoiv.net>
Co-authored-by: Jason R. Coombs <jaraco@jaraco.com>

Noltari committed:
Update Airzone Cloud to v0.5.2 and add fan speeds to Zones (#119314)

* Update aioairzone-cloud to v0.5.2
* airzone_cloud: climate: add zone fan speeds support

Signed-off-by: Álvaro Fernández Rojas <noltari@gmail.com>

emontnemery committed:
Fix statistic_during_period after core restart (#119323)

bdraco committed:
Migrate august to use yalexs 6.0.0 (#119321)

swcloudgenie committed:
Fix AladdinConnect OAuth domain (#119336)

fix aladdin connect oauth domain

jbouwh committed:
Fix enigma2 option flow (#119335)

bdraco committed:
Bump uiprotect to 0.10.1 (#119327)

Co-authored-by: Jan Bouwhuis <jbouwh@users.noreply.github.com>

bdraco committed:
Speed up generating large stats results (#119210)

* Speed up generating large stats results

* naming

* fix type

* fix type

* tweak

* tweak

* delete unused code

JakeMartin-ICL committed:
Move runtime_data deletion after unload (#119224)

* Move runtime_data deletion after unload.

Doing this before unload means we can't use, eg. the coordinator, during teardown.

* Re-order config entry on unload

* Add test

---------

Co-authored-by: Paulus Schoutsen <balloob@gmail.com>

bdraco committed:
Bump uiprotect to 0.13.0 (#119344)

elupus committed:
Allow source sensor to be changed in threshold helper (#119157)

* Allow source sensor to be changed in threshold helper

* Make sure old device link is removed on entry change

* Add test case for changed association

jbouwh committed:
Bump incomfort backend client to v0.6.2 (#119330)

middlingphys committed:
Fix typo in Ecovacs integration (#119346)

rubeecube committed:
Bump pyElectra to 1.2.1 (#118958)

Ishima committed:
Add Xiaomi Air Purifier Pro H EU (zhimi.airpurifier.vb2) (#119149)

jbouwh committed:
Customize incomfort binary sensor icons (#119331)

jesserockz committed:
Bump aioesphomeapi to 24.6.0 (#119348)

balloob committed:
OpenAI to respect custom conversation IDs (#119307)

epenet committed:
Avoid cross-domain imports in scrape tests (#119351)

epenet committed:
Use absolute import in roborock tests (#119353)

epenet committed:
Ignore some pylint errors in component tests (#119352)

epenet committed:
Fix namespace-import pylint warning in shelly tests (#119349)

kaareseras committed:
Fix Azure data explorer (#119089)

Co-authored-by: Robert Resch <robert@resch.dev>

bieniu committed:
Bump `imgw-pib` backend library to version 1.0.5 (#119360)

Co-authored-by: Maciej Bieniek <478555+bieniu@users.noreply.github.com>

gjohansson-ST committed:
Calculate attributes when entity information available in Group sensor (#119021)

Betacart committed:
Fix typo in Tibber service description (#119354)

epenet committed:
Fix pointless-string-statement pylint warning in emulated_hue tests (#119368)

epenet committed:
Ignore c-extension-no-member pylint warnings in tests (#119378)

emontnemery committed:
Allow importing typing helper in core files (#119377)

* Allow importing typing helper in core files

* Really fix the circular import

* Update test

emontnemery committed:
Fix typo in auth (#119388)

emontnemery committed:
Fix missing checks in core config test (#119387)

epenet committed:
Fix pylint warnings in xiaomi tests (#119373)

epenet committed:
Fix import-outside-toplevel pylint warnings in tests (#119389)

epenet committed:
Fix pylint warnings in testing config custom components (#119370)

epenet committed:
Fix non-parent-init-called pylint warning in google_assistant tests (#119367)

epenet committed:
Fix typo in pylint plugin (#119362)

epenet committed:
Improve type hints in yaml util tests (#119358)

epenet committed:
Fix confusing-with-statement pylint warnings (#119364)

epenet committed:
Fix import pylint warning in core tests (#119359)

jafar-atili committed:
Bump pyElectra to 1.2.3 (#119369)

epenet committed:
Fix use-maxsplit-arg pylint warnings in tests (#119366)

epenet committed:
Fix consider-using-with pylint warnings in matrix tests (#119365)

epenet committed:
Improve type hints and fix pylint warnings in util tests (#119355)

emontnemery committed:
Use statistic tables' duration attribute instead of magic numbers (#119356)

Sese-Schneider committed:
Add switch to Tuya thermostat: child_lock (#113052)

epenet committed:
Fix incorrect constants in streamlabswater tests (#119399)

epenet committed:
Drop use of deprecated constant in zha tests (#119397)

epenet committed:
Fix incorrect constants in google_travel_time tests (#119395)

epenet committed:
Fix import-outside-toplevel pylint warnings in core tests (#119394)

epenet committed:
Remove deprecated imports in config tests (#119393)

epenet committed:
Fix unnecessary-dunder-call pylint warnings in tests (#119379)

emontnemery committed:
Support shared keys starting with period in services.yaml (#118789)

epenet committed:
Ignore no-name-in-module warnings in tests (#119401)

epenet committed:
Fix pointless-exception-statement warning in tests (#119402)

epenet committed:
Fix use-implicit-booleaness-not-len warnings in tests (#119407)

epenet committed:
Fix unspecified-encoding warnings in tests (#119405)

epenet committed:
Ignore misplaced-bare-raise warnings in tests (#119403)

epenet committed:
Adjust nacl import in tests (#119392)

timmo001 committed:
Restructure and setup dedicated coordinator for Azure DevOps (#119199)

dkrahmer committed:
Add support for Tuya non-standard contact sensors (#115557)

Co-authored-by: Franck Nijhof <git@frenck.dev>

Kane610 committed:
Merge branch 'kimdv/add-support-for-number-of-clients-attached-to-a-device' of https://github.com/kimdv/core into pr/117207

Kane610 commented:
New PR is up to side step my rebase debacle :)
